subprojects/core-api/src/main/java/org/gradle/api/file/FileContents.java

```diff
@@ -47,6 +50,24 @@ public interface FileContents {
      */
     Provider<String> getAsText();
 
+    /**
+     * Gets a provider of the entire file contents as a single String.
+     *
+     * See {@link #getAsText()} for details
+     *
+     * @param charset Charset to be used
+     * @return provider of the entire file contents as a single String.
+     * @since 8.7
+     */
+    @Incubating
+    default Provider<String> getAsText(String charset) {
+        if (Charset.defaultCharset().toString().equals(charset)) {
+            return getAsText();
+        } else {
```

Comment:
❓ 

I'd lose the `else` bit, maybe inverting the condition.

But more importantly, is this an interface we expect or know folks to implement? If not, do we need the default implementation?

Reply:
> But more importantly, is this an interface we expect or know folks to implement? If not, do we need the default implementation?

I still do not know Gradle's rules and practices, so I used Jenkins' approach when we assume every public API is public unless annotated or documented otherwise. Here, there are no docs though @ljacomet also thinks we should be able to do a breaking interface change here

Reply:
> > But more importantly, is this an interface we expect or know folks to implement? If not, do we need the default implementation?
> 
> I still do not know Gradle's rules and practices, so I used Jenkins' approach when we assume every public API is public unless annotated or documented otherwise. Here, there are no docs though @ljacomet also thinks we should be able to do a breaking interface change here

:thinking: instead of throwing an exception, you can use something like:
```java
getAsBytes().map(transformer(bytes -> new String(bytes, charset)))
```
Which is probably slightly less efficient, but works everywhere. Given that the result is in-memory anyway, we're not talking about multi-GB strings there.